synapse/lib/cell.py

```diff
@@ -1486,36 +1492,46 @@ async def _runSysctlLoop(self):
 
             await self.waitfini(self.SYSCTL_CHECK_FREQ)
 
-    def _getAhaAdmin(self):
-        name = self.conf.get('aha:admin')
-        if name is not None:
-            return name
-
     async def _initAhaRegistry(self):
 
-        ahaurl = self.conf.get('aha:registry')
-        if ahaurl is not None:
+        ahaurls = self.conf.get('aha:registry')
+        if ahaurls is not None:
 
-            info = await s_telepath.addAhaUrl(ahaurl)
-            if self.ahaclient is None:
-                self.ahaclient = await s_telepath.Client.anit(info.get('url'))
-                self.ahaclient._fini_atexit = True
-                self.onfini(self.ahaclient)
+            info = await s_telepath.addAhaUrl(ahaurls)
+            if self.ahaclient is not None:
+                await self.ahaclient.fini()
 
-            async def finiaha():
-                await s_telepath.delAhaUrl(ahaurl)
+            async def onlink(proxy):
+                ahauser = self.conf.get('aha:user', 'root')
+                newurls = await proxy.getAhaUrls(user=ahauser)
+                oldurls = tuple(self.conf.get('aha:registry'))
+                if newurls and newurls != oldurls:
+                    self.modCellConf({'aha:registry': newurls})
+                    self.ahaclient.setBootUrls(newurls)
 
-            self.onfini(finiaha)
+            self.ahaclient = await s_telepath.Client.anit(ahaurls, onlink=onlink)
+            self.onfini(self.ahaclient)
 
-        ahauser = self.conf.get('aha:user')
-        ahanetw = self.conf.get('aha:network')
+            async def fini():
+                await s_telepath.delAhaUrl(ahaurls)
+
+            self.ahaclient.onfini(fini)
 
-        ahaadmin = self._getAhaAdmin()
+        ahaadmin = self.conf.get('aha:admin')
         if ahaadmin is not None:
             await self._addAdminUser(ahaadmin)
 
-        if ahauser is not None:
-            await self._addAdminUser(ahauser)
```

Comment:
If we remove this, we'll be breaking deployments where users have specified the `aha:user` in their provisioning configurations ( ex. `docker exec -it 00aha-aha-1 python -m synapse.tools.aha.provision.service --user synapse --only-url 02.cortex` ), since the target service would not have the `aha:user` created to auth against with the client certificate. This is something that we have deployed in the wild.

Reply:
The `aha:user` is used to determine the username when generating AHA URLs not when connecting to the service. Am i forgetting something?

Reply:
The `aha:user` is used in a few places in `master`; but notably when we create the provisioning information a service, we use it in a few places:

1. We create a user in the AHA cell for that user value and grant it permission to register itself  and get services.
2. We mod the `aha:user` into the `user` field of the `aha:registry` URLs given to a service, so its implicitly used to connect to AHA.

Then on the Cell during provisioning, we have AHA sign the `aha:user` cert and later make the `aha:user` user.

Later on, whenever a telepath Client calls `getAhaSvc()`, the AHA service inserts the link username ( which in the service to service model is `aha:user` ) we modify the urlinfo to include the link username. This means a service connecting from SVC A to SVC B uses that `aha:user` value.

So, this means we need to 1) update our `aha:registry` settings in the initial configuration to add the `aha:user` value and 2) update the Cell call that gets the `aha:registry` to include the `aha:user` information.